src/nvim/api/ui.c

```diff
@@ -112,6 +114,32 @@ void nvim_ui_attach(uint64_t channel_id, Integer width, Integer height,
 
   pmap_put(uint64_t)(connected_uis, channel_id, ui);
   ui_attach_impl(ui);
+
+  dict_T *dict = get_vim_var_dict(VV_EVENT);
+  tv_dict_add_nr(dict, S_LEN("channel_id"), (long)channel_id);
+  tv_dict_add_nr(dict, S_LEN("width"), (long)width);
+  tv_dict_add_nr(dict, S_LEN("height"), (long)height);
+
+  dict_T *const options_vim = tv_dict_alloc();
+  for (uint32_t i = 0; i < options.size; i++) {
+    KeyValuePair item = options.items[i];
+    String key = item.key;
+
+    if (key.size == 0) {
+      continue;
+    }
+
+    dictitem_T *const di = tv_dict_item_alloc(key.data);
+    if (!object_to_vim(item.value, &di->di_tv, err)) {
+      tv_dict_item_free(di);
+    }
+    tv_dict_add(options_vim, di);
+  }
+
+  tv_dict_add_dict(dict, S_LEN("options"), options_vim);
+  tv_dict_set_keys_readonly(dict);
+  apply_autocmds(EVENT_UIATTACH, NULL, NULL, false, curbuf);
```

Comment:
`apply_autocmds` should be queued on the main loop, rather than immediately executed.

    multiqueue_put(main_loop.events ...

Reply:
No, this is already on the main loop and safe.

Reply:
I'm not concerned about threads. when new events are added, executing VimL in new codepaths is often buggy. It happened with WinNew. There's no reason to execute this code immediately here, it's better to defer to the next tick.

Reply:
I never said anything about threads. This _already_ is in a handler of a _deferred_ event. Absolutely nothing is gaining by "deferring to the next tick" because we already is one tick deferred from whatever dangerous place you want to avoid.

Reply:
Rather than safety, I wonder if we should avoid the risk of slow plugins potentially slowing-down UI attach.